python/ray/data/tests/test_optimize.py

```diff
@@ -87,11 +87,13 @@ def test_spread_hint_inherit(ray_start_regular_shared):
     ds = ray.data.range(10)._experimental_lazy()
     ds = ds.map(lambda x: x + 1)
     ds = ds.random_shuffle()
-    for s in ds._plan._stages:
+    for s in ds._plan._stages_before_snapshot:
```

Comment:
It would be great to have more tests for this unlinking behavior as it's fairly intricate.